src/Framework/Sdk/SdkResultItem.cs

```diff
@@ -34,35 +40,24 @@ public override bool Equals(object obj)
         {
             if (obj is SdkResultItem item &&
                    ItemSpec == item.ItemSpec &&
-                   Metadata?.Count == item.Metadata?.Count)
+                   item.Metadata is not null &&
+                   Metadata?.Count == item.Metadata.Count)
             {
-                if (Metadata != null)
-                {
-                    foreach (var kvp in Metadata)
-                    {
-                        if (item.Metadata[kvp.Key] != kvp.Value)
-                        {
-                            return false;
-                        }
-                    }
-                }
-
-                return true;
+                return Metadata.All(m => item.Metadata.TryGetValue(m.Key, out var itemValue) && itemValue == m.Value);
             }
             return false;
         }
 
         public override int GetHashCode()
         {
             int hashCode = -849885975;
-            hashCode = (hashCode * -1521134295) + EqualityComparer<string>.Default.GetHashCode(ItemSpec);
+            hashCode = hashCode ^ ItemSpec.GetHashCode();
 
             if (Metadata != null)
             {
                 foreach (var kvp in Metadata)
                 {
-                    hashCode = (hashCode * -1521134295) + kvp.Key.GetHashCode();
-                    hashCode = (hashCode * -1521134295) + kvp.Value.GetHashCode();
+                    hashCode ^= $"{kvp.Key}: {kvp.Value ?? "V"}".GetHashCode();
```

Comment:
@rainersigwald 
Could I use StringBuilder to minimize allocations?

Reply:
You could probably change this to hashCode ^= kvp.Key.GetHashCode() * ((kvp.Value ?? "V").GetHashCode() + 1); and avoid allocations while ignoring order, maintaining association importance, and differentiating between keys and values.

Reply:
`kvp.Key.GetHashCode()` is wrong. Because the metadata dictionary uses OrdinalIgnoreCase comparison for the keys, GetHashCode() must do the same; please see <https://github.com/dotnet/msbuild/issues/6750#issuecomment-898233796>.

Does this project already reference an assembly that provides [HashCode.Combine](https://docs.microsoft.com/dotnet/api/system.hashcode.combine?view=dotnet-plat-ext-3.1#System_HashCode_Combine__2___0___1_)?

Reply:
You're right about the case sensitivity weirdness, and it would be good to fix that, but it wasn't correct before, and eriawan said they planned just to resolve the first few bullets, so I'm assuming we're leaving that for a later PR. Is that correct, @eriawan?

No, I don't think so, though I agree it would be nice here.

Reply:
@Forgind 

Yes, my initial plan was to leave the case sensitivity for later PR.
Do you want that to be solved too?
We still have allocation problem that needs to be solved, as suggested by @rainersigwald

Reply:
Apparently you can just say OrdinalIgnoreCase.GetHashCode(<string>), and it will get you a case-insensitive hash.

> hashCode ^= kvp.Key.GetHashCode() * ((kvp.Value ?? "V").GetHashCode() + 1);

This was my suggestion to get away from the allocation.

Reply:
@Forgind 

branch rebased to main, and GetHashCode updated. 
There were failing unit tests, but those failed tests aren't related to my updates at all.
Please review, thanks in advance! 🙏